### PR TITLE
Backend chats refactor

### PR DESCRIPTION
**Pull request by matt989253**

### CODE REVIEW

File: src/web/src/main/java/tupperdate/web/facade/chats/ChatFacadeImpl.kt [hunk 1]
```diff
@@ -1,38 +1,146 @@
...
+            val chatUser =
+                when (val res = userRepository.read(user = User(AccountId(conv.theirId)))) {
+                    is Result.Ok -> res.result
+                    else -> return res.map { listOf() }
+                }
+            val lastMessage =
+                when (val res = chatRepository.readMessages(user = user, userId = conv.theirId)) {
+                    is Result.Ok -> res.result
+                    else -> return res.map { listOf() }
+                }.getOrNull(0)
```
Comment (alexandrepiveteau) [inline src/web/src/main/java/tupperdate/web/facade/chats/ChatFacadeImpl.kt]: We probably don't want to scan through the whole message set of a user each time, but rather only take the latest directly.

File: src/web/src/main/java/tupperdate/web/facade/chats/ChatFacadeImpl.kt [hunk 2]
```diff
@@ -1,38 +1,146 @@
...
+
+        val chatUser = when (val res = userRepository.read(user = User(AccountId(conv.theirId)))) {
+            is Result.Ok -> res.result
+            else -> return res.map { emptyChat() }
+        }
+        val lastMessage =
+            when (val res = chatRepository.readMessages(user = user, userId = conv.theirId)) {
+                is Result.Ok -> res.result
+                else -> return res.map { emptyChat() }
+            }.getOrNull(0)
```
Comment (alexandrepiveteau) [inline src/web/src/main/java/tupperdate/web/facade/chats/ChatFacadeImpl.kt]: We probably don't want to scan through the whole message set of a user each time, but rather only take the latest directly.

File: src/web/src/main/java/tupperdate/web/facade/chats/ChatFacadeImpl.kt [hunk 3]
```diff
@@ -1,38 +1,146 @@
...
+        )
+
+        // Let all the two clients sync their chats.
+        if (result is Result.Ok) {
+            notificationRepository.dispatch(
+                notification = arrayOf(
+                    Notification.ToUser.UserSyncOneConversation(userId, user.id.uid),
+                    Notification.ToUser.UserSyncOneConversation(user.id.uid, userId),
+                )
+            )
```
Comment (alexandrepiveteau) [inline src/web/src/main/java/tupperdate/web/facade/chats/ChatFacadeImpl.kt]:
```suggestion
            notificationRepository.dispatch(
                Notification.ToUser.UserSyncOneConversation(userId, user.id.uid),
                Notification.ToUser.UserSyncOneConversation(user.id.uid, userId),
            )
```

File: src/web/src/main/java/tupperdate/web/model/chats/Conversation.kt [hunk 4]
```diff
@@ -0,0 +1,9 @@
+package tupperdate.web.model.chats
+
+data class Conversation(
+    val id: String,
+    val myId: String,
+    val theirId: String,
+    val myRecipes: List<String>,
+    val theirRecipes: List<String>,
```
Comment (alexandrepiveteau) [inline src/web/src/main/java/tupperdate/web/model/chats/Conversation.kt:8]: Use `PictureUrl` here.

File: src/web/src/test/java/tupperdate/web/chats.kt [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+            // Put user profile
+            handleRequest(HttpMethod.Put, "/users/$botId") {
+                authRequest(botId)
+                jsonType()
+                val body = MyUserDTO(displayName = botName, imageBase64 = null)
+                setBody(Json.encodeToString(body))
+            }.apply { assertEquals(HttpStatusCode.OK, response.status()) }
+
+            // Get chat list
+            handleRequest(HttpMethod.Get, "/recipes/own") {
```
Comment (alexandrepiveteau) [inline src/web/src/test/java/tupperdate/web/chats.kt]: Bad endpoint.

File: src/web/src/main/java/tupperdate/web/model/chats/Conversation.kt (hunk 4 above)
Comment (matt989253) [reply src/web/src/main/java/tupperdate/web/model/chats/Conversation.kt:8]: These are recipe ids, not pictures. Care to explain in more details?

File: src/web/src/test/java/tupperdate/web/chats.kt [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+            assertEquals(bot1MessageContent.tempId, messagesBot1[0].tempId)
+            assertEquals(bot2MessageContent.tempId, messagesBot2[1].tempId)
+        }
+    }
+
+    private fun TestApplicationEngine.getConversation(
+        botId: String,
+        userId: String,
+    ): ConversationDTO {
+        handleRequest(HttpMethod.Get, "/chats/$userId/messages") {
```
Comment (alexandrepiveteau) [inline src/web/src/test/java/tupperdate/web/chats.kt]: I'm not sure about this.

**Review by alexandrepiveteau — APPROVED** [review]

**Pull request merged**